### ISSUE TITLE
Update Build Images

### ISSUE THREAD

**Issue by DanRStevens** [lairworks/nas2d-core#1155]
The build images haven't been updated in awhile and should be updated.

---

The next LTS (Long Term Support) version of Ubuntu is out this month.
https://releases.ubuntu.com/noble/
- `24.04` (Noble Numbat)

----

Additionally, the MacOS build image is now quite dated.
https://circleci.com/docs/using-macos/#supported-xcode-versions
- Most recent documented: `15.3.0` (marked as unsupported by Visual Studio code)
- Most recent in file format schema: `15.2.0` (accepted without error underline squiggles)
- Currently using: `13.4.1`

**Comment by DanRStevens** [comment]
PR #1166 Added a workflow to build Docker images and push them to the GitHub Docker registry. Though it may trigger runs more often then needed.

For instance, if a Docker image build is done on a branch, and then tested by building the code using the new Docker image, and then that branch is merged to `main`, it will trigger another build of the Docker image. We should probably avoid that since it could potentially produce a different result from what the branch built (if some software update became available during the meantime), and so future builds would run with the different build that wasn't the one tested and verified.

It might make sense to not automatically run the new workflow for the `main` branch, since there is an assumption that Docker image updates will be tested and verified on a branch, and then future builds using the updated version would use the version built while testing and developing on the branch.

Potentially we could add a `push` `paths` filter of `!main`, and a comment explaining why.

----

Another weakness if that if we were to fill out the `matrix` to do builds for other Linux distributions, they would all build together, even if only one of them had an updated Dockerfile or version number. The limitation here is that the workflow as a whole can be triggered using a path filter, but jobs are not given such an option. If they were, we could have filtered on `paths` such as `dockerfile/nas2d-${{ matrix.platform }}.*`. Sadly, such as path filter can only be used on the workflow, and so needs to apply to all Dockerfiles.

Given the limitations of path filters, we can't prevent a matrix job from running, though we could have a job terminate early, such as by detecting if a Docker image has already been built.

The `docker` command line tool has a new experimental feature which can determine if an image and tag already exist in a remote repository, reporting the result using the exit code from the command. Details:
- https://stackoverflow.com/questions/32113330/check-if-imagetag-combination-already-exists-on-docker-hub
> `docker manifest inspect $IMGNAME:$IMGTAG > /dev/null ; echo $?`

One downside to that, is that while working on a branch to test and verify a new build, it's expected the Docker image will be overwritten with each new build. We don't expect the version number to be continually bumped for experimental builds. It's only when the branch is merged that things should be finalised.

**Comment by DanRStevens** [comment]
Running `git diff` the `--exit-code` flag will return `1` if there were differences and `0` if there were no differences. This would be even easier than listing just the filenames of changed files and scanning them for specific files.

Using `--exit-code`, it may also make sense to replace `--name-only` with `--no-patch`/`-s` instead, which suppresses output. If all we need is the return code, then there is no need to print the filenames of changed files. Though perhaps it may still be useful for debugging or inspection purposes to print filenames of changed files.

StackOverflow: [Check if specific file in git repository has changed](https://stackoverflow.com/questions/17797740/check-if-specific-file-in-git-repository-has-changed)

Example usages:
- `git diff --exit-code --name-only main -- docker/nas2d-arch.*; echo $?`
- `git diff --exit-code --no-patch main -- docker/nas2d-arch.*; echo $?`

**Comment by DanRStevens** [comment]
A potential problem with running `git diff` on a runner, is the default checkout mode uses `fetch-depth: 1`, so there won't be any other commits to compare with. Additionally, turning on the `--depth` parameter to a `git fetch` also enables `--single-branch` by default, so there won't be any information about other branches.

To add information for a single branch, such as branch `main` from `origin`:
```
git remote set-branches --add origin main
```

The commit for that branch would then need to be fetched, perhaps also using a `fetch-depth` of 1:
```
git fetch --depth 1
```

At that point, a comparison should then be possible:
```
git diff --name-only origin/main -- docker/
```

Information on the state of the repo (throughout the process above) can be seen with:
```
git remote show origin
```

Additionally, `fetch` information can be seen in `.git/config`.

The commands and environment can be tested locally by cloning a repo using a `--depth 1` setting:
```
git clone --depth 1 file:///path/to/repo/nas2d-core/ clone-test
```

----

Alternatively, instead of adding a remote tracking branch, the branch can be downloaded into a local branch. Doing so is about the same for a CI environment, though perhaps less convenient long term if working locally, where updates may need to be fetched at some point in the future.

Fetching commit directly into a local branch:
```
git fetch --depth 1 origin main:main
```

Since the branch is local, the `git diff` needs to be slightly different (no longer using `origin/main`, but just `main`):
```
git diff --name-only main -- docker/
```

**Comment by DanRStevens** [comment]
The Mingw Docker environment shows warnings when installing the `apt` repo from `wine`.

> Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).

> W: https://dl.winehq.org/wine-builds/ubuntu/dists/noble/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.

This stems from the use of `apt-key` in `apt-key add -`, which is now a deprecated tool.

----

A lot of examples on the internet give an alternative using `gpg --dearmor`, however, they often given instructions to place keys files in the same place that `apt-key` used to place them, which may be part of why it was deprecated.

Some additional info about why not to place keys in the often suggested location:
- https://askubuntu.com/questions/1286545/what-commands-exactly-should-replace-the-deprecated-apt-key
- https://unix.stackexchange.com/questions/463072/why-does-the-apt-key-man-page-advise-against-using-its-add-command

In summary, avoid using `/etc/apt/trusted.gpg.d/` when installing keys, and instead use a different folder, such as `/etc/apt/keyrings/`.

----

Some additional info can be found in:
- `man sources.list`
    -  section `signed-by`
        - `/usr/share/keyrings/` (for keyrings managed by packages)
        - `/etc/apt/keyrings/` (for keyrings managed by the system operator)
> The recommended locations for keyrings are `/usr/share/keyrings` for keyrings managed by packages, and `/etc/apt/keyrings` for keyrings managed by the system operator.

**Comment by DanRStevens** [comment]
Worked on updating the Mingw build image, though currently it's unable to run the unit test executable:
> wine: failed to open "../.build/Debug_Linux_test/test": c0000135

Explicitly specifying the `wine64` executable seems to allow it to run:
`cd test && /usr/lib/wine/wine64 ../.build/Debug_Linux_test/test.exe`

Oddly, when running the image as `root` to debug the issue, using just `wine` appeared to run the unit test executable correctly, but the same command as a more limited user would fail. Only explicitly choosing the `wine64` binary would work as a normal user.

----

The `wine` command is picked up as a series of symlinks:
`/usr/bin/wine` -> `/etc/alternatives/wine` -> `/usr/bin/wine-stable`

It appears `wine-stable` is a shell script, which selects between the 32-bit or the 64-bit version, which are found at:
- `wine32=/usr/lib/wine/wine`
- `wine64=/usr/lib/wine/wine64`

From the script, it appears the 32-bit version is preferred, when it is available. When it's not available, a message is printed, suggesting it be installed, with instructions on how to install the 32-bit version.

**Comment by DanRStevens** [comment]
When updating the Mingw environment to Ubuntu 24.04, there are updated packages for both Mingw and Wine. The Mingw build completes fine, but the resulting executable terminates without normal run output and without an error message. From this it's unclear if there is a problem with Wine, or with the executable produced by the newer version of Mingw.

If that same binary is run in an older version of the build image (using an older version of Wine), it runs partially, but aborts the unit test run with an unhandled page fault. The last run test was `Sprite.animationCompleteSignal`. This would seem to suggest there may be a problem with the produced executable, rather than Wine.

The failures happen when running the test for `Sprite.animationCompleteSignal`. This leads me to suspect the executable compiled in the newer environment is hitting problems with the undefined behavior mentioned in:
- Issue #1159

**Comment by DanRStevens** [comment]
It seems fixing the `gmock` undefined behavior issue did not solve the problem mentioned above.

Compiling a unit test executable in an updated Mingw environment on Ubuntu 24.04, and then attempting to run that executable on the existing Ubuntu 22.04 image produces errors in a number of the unit tests.

The following command was used to filter out certain tests and get the remaining tests to run:
```
wine ../.build/Debug_Linux_test/test.exe --gtest_filter="-SpriteCompleteSignal.animationCompleteSignalNone:SpriteCompleteSignal.animationCompleteSignalOnceDelta2:SpriteCompleteSignal.animationCompleteSignalOnceDelta3:SpriteCompleteSignal.animationCompleteSignalTwice:SpriteCompleteSignal.animationCompleteSignalStopFrameDelta0:SpriteCompleteSignal.animationCompleteSignalStopFrameDelta4:Version.versionString:Filesystem.searchPath:Filesystem.directoryList:Filesystem.mountUnmount:SignalConnection.Scope:Delegate.DelegateCall:Delegate.DelegateCallConst:Signal.ConnectEmitDisconnect"
```

The problem tests are:
- `Delegate.DelegateCall`
- `Delegate.DelegateCallConst`
- `Signal.ConnectEmitDisconnect`
- `SignalConnection.Scope`
- `SpriteCompleteSignal.animationCompleteSignalNone`
- `SpriteCompleteSignal.animationCompleteSignalOnceDelta2`
- `SpriteCompleteSignal.animationCompleteSignalOnceDelta3`
- `SpriteCompleteSignal.animationCompleteSignalTwice`
- `SpriteCompleteSignal.animationCompleteSignalStopFrameDelta0`
- `SpriteCompleteSignal.animationCompleteSignalStopFrameDelta4`
- `Version.versionString` (`MatchesRegex`)
- `Filesystem.searchPath`
- `Filesystem.directoryList`
- `Filesystem.mountUnmount`

The problems seem to fall into 3 categories:
- Use of `Delegate`
- Use of `MatchesRegex`
- Use of certain `Filesystem` functions

**Comment by DanRStevens** [comment]
Sticking with Ubuntu 22.04, it's possible to upgrade the version of Wine from `8.0.2` to `9.0.0.0`. This keeps the version of Mingw constant. By compiling with a fixed version of Mingw, and showing the resulting binary runs properly with Wine `8.0.2` but not with Wine `9.0.0.0` suggests there was a change with Wine that broke the unit test runs.

Perhaps [`winedbg`](https://linux.die.net/man/1/winedbg) could be used to help investigate the cause of the problem.

Alternatively, it may be possible to use remote `gdb` debugging.

[How to properly debug a cross-compiled Windows code on linux?](https://stackoverflow.com/questions/39938253/how-to-properly-debug-a-cross-compiled-windows-code-on-linux)
> `apt install gdb-mingw-w64 gdb-mingw-w64-target`

**Comment by DanRStevens** [comment]
Testing with the Docker image containing Wine version 9, it's possible to compile and run a simple executable and get debug output using `std::cout`.

Testing with a modified unit test project that added a `main.cpp` with simple `std::cout` output, the following warning was seen, with no run output:
> warning: no previous declaration for 'int SDL_main()' [-Wmissing-declarations]
> test/main.cpp:4:5: note: in expansion of macro 'main'

Looking at the flags beings passed to the compiler and linker, it appears the `-lSDL_main` flag is being passed during linking. That flags was added by `sdl2-config --libs`. There is no command line option to suppress output of that flag. Hacking the `makefile` a bit to avoid calling `sdl2-config`, while retaining other flags, it's possible to compile the unit test project, and get full unit test output.

It appears the problem of no output was `SDL_main` hijacking the `main` function.

I'm kind of thinking perhaps using `sdl2-config` is not really a best practice. It's starting to seem like a hacked together way specific to the SDL project. There's probably a reason why it was not replicated elsewhere.

----

It's not clear why the was able to run with proper output on the older image version, but produces no output on the newer image version.

There were no differences in link flags between the two versions (images with Wine 8 vs Wine 9):
```sh
x86_64-w64-mingw32-g++ <*.o> lib/libnas2d.a  -lgmock_main -lgmock -lgtest -lpthread  -lstdc++ -lSDL2_image -lSDL2_mixer -lSDL2_ttf -L/usr/local/x86_64-w64-mingw32/lib -lmingw32 -mwindows -lSDL2main -lSDL2 -lglew32 -lopengl32 -o .build/Debug_Linux_test/test.exe
```

Checking a sample of the compile flags also produced no differences:
```sh
x86_64-w64-mingw32-g++ -MMD -MP -DGLEW_STATIC -I./  -Og -g -std=c++20 -Wall -Wextra -Wpedantic -Wzero-as-null-pointer-constant -Wnull-dereference -Wold-style-cast -Wcast-qual -Wcast-align -Wdouble-promotion -Wshadow -Wnon-virtual-dtor -Woverloaded-virtual -Wmissing-declarations -Wmissing-include-dirs -Winvalid-pch -Wmissing-format-attribute -Wredundant-decls -Wformat=2 -Wno-redundant-decls -I/usr/local/x86_64-w64-mingw32/include/SDL2 -Dmain=SDL_main  -c -o .build/Debug_Linux_test/intermediate/Math/Vector.test.o test/Math/Vector.test.cpp
```

Actually, compiling on the older version produces output in the older environment, yet running it again in the newer environment without a recompile produces no output. This may suggest a different in runtime linking.
